### ISSUE TITLE
Waveform freezes when repeatedly pressing CUE

### ISSUE THREAD

**Issue by Holzhaus** [Holzhaus/mixxx-gh-issue-migration#15]
Using trunk on Debian testing with ALSA, Celeron D 2.53GHz.

Steps to reproduce:

1) Load the attached song (happens with many songs)
2) Set a cue point at around 4s (where she says "Ohh" in this song)
3) Press CUE to play from the cue point
4) Release & press it again & notice that the waveform doesn't respond. Sometimes there is sound, sometimes not.
5) Wait 5 seconds or so then try it again & it'll work, until you release & press CUE again.
Also doesn't respond to PLAY, FF, REW, etc. Can also cause it by dragging the waveform backwards across the cue point.

------------------------------------
Imported from Launchpad using lp2gh.

 * date created: 2009-03-13T04:47:09Z
 * owner: pegasus-renegadetech
 * the launchpad url was https://bugs.launchpad.net/bugs/342113

**Comment by Holzhaus** [comment]
(by pegasus-renegadetech)

**Comment by Holzhaus** [comment]
(by pegasus-renegadetech)
Assigning to rryan since he's da waveform masta.
This may be related to or a duplicate of bug #257457.

**Comment by Holzhaus** [comment]
(by gamegod)
Does the music stop? I think it might be related to the deadlock I saw
yesterday.

I'm pretty sure 257457 is dead...

**Comment by Holzhaus** [comment]
(by pegasus-renegadetech)
Sometimes the music stops, sometimes not. It's hard to tell since I'm bashing CUE. I have seen it continue while the waveform is stuck if I hit play, though.

**Comment by Holzhaus** [comment]
(by pegasus-renegadetech)
This is also a problem in 1.6.1, FWIW, though it recovers much quicker and the music doesn't stop.

**Comment by Holzhaus** [comment]
(by rryan)
Does this occur while machine-gunning the hotcues on the SCS3d?

**Comment by Holzhaus** [comment]
(by pegasus-renegadetech)
Yes but also when just using the CUE button in the GUI with no MIDI devices connected.

**Comment by Holzhaus** [comment]
(by rryan)
I can't seem to reproduce this with the release branch. I remember this being an issue though.

**Comment by Holzhaus** [comment]
(by rryan)
With Sean's help I was able to reproduce this. I tracked it down to EngineBuffer::process(). 

First some background: Every time through process(), various control objects are updated. Some are done on each call (visual_playposition), and some are done after every few calls (playposition, rate, etc). These updates are inside of an if block which first decides whether or not to wake up the Reader to tell it to read more samples from the SoundSource. The if statement tests whether or not the EngineBuffer was able to acquire a lock on the Reader this time through process(). I'm not sure why the updates were placed in this block in the first place, but this logic is wrong. 

The problem that was causing the freeze is that the MP3 that Sean posted has a frame error at around :02. This particular error makes libmad hang for a relatively long while. Since SoundSourceMP3::read() was hanging, the Reader was hanging in a locked state, so the EngineBuffer could not lock the Reader to sync the buffer details (a step it takes each time process() is called). The if statement (more or less) tests whether or not the EngineBuffer was able to lock the Reader on this cycle. So when the frame error happens, the visual_playposition ControlObject is not updated for about 300ms while libmad is blocking. This causes the waveform to get stuck, but the audio to keep playing (since the Reader has a few seconds of the song data buffered). 

I just committed a fix to this to the release branch, r2331. The fix simply pulls the ControlObject update code out of the if statement testing for whether the Reader was locked.
